### ISSUE TITLE
quick-add for mandatory fields

### ISSUE THREAD

**Issue by tsohst** [netbox-community/netbox#5858]
### Environment
* Python version:  2.7.16
* NetBox version: v2.10.3

### Proposed Functionality
When adding a new device for example, you always have the finish creating anything related (mandatory fields).
A "+" button would be useful, which would direct you the the specific editor (device-role in this case), you can then create the device-role - save and netbox redirects you to the original (device) editor and auto-fill the field.

![image](https://user-images.githubusercontent.com/32710648/109009111-6e30a180-76ae-11eb-82b7-02b97622a5f7.png)

### Use Case
Faster creating of anything that has relations (mandatory field) (e.g. devices -> device-roles)

### Database Changes
Don't know

### External Dependencies
Don't know

**Comment by jeremystretch** [comment]
> which would direct you the the specific editor

We could add an "add" button to the dropdown easily, but are you saying have it open a separate tab, or something more advanced? Ideally it would open a modal within the current tab to add the new object, but that would take a lot more work under the hood to implement.

**Comment by tsohst** [comment]
No new Tab, open the editor in the same Tab, to create the new object.
Save that new object  and after save, your back in the editor you we're before.

**Comment by jeremystretch** [comment]
Yeah, as I mentioned that's not going to be feasible without a huge amount of UI work unfortunately.

**Comment by DanSheps** [comment]
I have to concur with @jeremystretch, until we do a major UI rewrite, modals are just not feasible right now.

**Comment by jeremystretch** [comment]
I think it's worthwhile to add a button though. Even if you need to open a new tab to add something, the selection widget within the form will automatically repopulate to include it.

**Comment by tsohst** [comment]
Okay, sounds good 👍

**Comment by cpmills1975** [comment]
Would it be possible to leverage the return_url somehow to bring the user back to the form they were editing after clicking save?

**Comment by DanSheps** [comment]
We could perhaps see how it is opened (new tab) and maybe do a window(tab) close instead

**Comment by jeremystretch** [comment]
Marking this as blocked by #5893

**Comment by jeremystretch** [comment]
Lately I've been looking into [HTMX](https://htmx.org/) and this seems like a prime use case for it. It should be pretty simple to embed an "add" button next to the field which, when clicked, displays an object creation form directly below the form field. Once the form is submitted, it will disappear, and the newly created object will appear as an available option in the field's dropdown widget.

**Comment by jeremystretch** [comment]
I've started working on a POC for this. The bare functionality is there, but we need to fix up some of the frontend stuff. Specifically:

- Applying APISelect to the modal fields in the "add new" form
- Fixing auto-slugification
- Rendering the success/error messages returned when submitting the form
- Automatically refreshing the list of available objects & selecting the newest one (if feasible)

**Comment by tsohst** [comment]
Hello @jeremystretch 

are there any updates on this issue?
I saw that all related issues are already closed?

**Comment by jeremystretch** [comment]
Hoping to take a fresh look at this soon.

**Comment by jeremystretch** [comment]
Bumped this to v4.0 as it probably makes more sense to implement as part of the UI refresh.

**Comment by jeffgdotorg** [comment]
Let's give this FR some weight in v4.1 planning. It addresses a family of paper-cut issues that we've been hearing a lot about lately.
